tests/python/unittest/test_operator.py

```diff
@@ -1619,6 +1620,193 @@ def test_sequence_mask():
+                    y_displacement = grid[i,yout,xout,1]
+
+                    xcoord = x_displacement + xout
+                    ycoord = y_displacement + yout
+
+                    xInTopLeft = int(floor(xcoord))
+                    xWeightTopLeft = max(0, 1-abs(xcoord - xInTopLeft))
+
+                    yInTopLeft = int(floor(ycoord))
+                    yWeightTopLeft = max(0, 1-abs(ycoord - yInTopLeft))
+
+                    TopLefIsIn = xInTopLeft >= 0 and xInTopLeft <= width-1 and yInTopLeft >= 0 and yInTopLeft <= height-1;
+                    TopRightIsIn = xInTopLeft + 1>= 0 and xInTopLeft + 1 <= width-1 and yInTopLeft >= 0 and yInTopLeft <= height-1;
+                    BottomLeftIsIn = xInTopLeft >= 0 and xInTopLeft <= width-1 and yInTopLeft+1 >= 0 and yInTopLeft+1 <= height-1;
+                    BottomRightIsIn = xInTopLeft+1 >= 0 and xInTopLeft+1 <= width-1 and yInTopLeft+1 >= 0 and yInTopLeft+1 <= height-1;
+
+                    inTopLeft = 0
+                    inTopRight = 0
+                    inBottomLeft = 0
+                    inBottomRight = 0
+
+                    for channel in range(num_channel):
+
+                        if (TopLefIsIn): inTopLeft = data[i,yInTopLeft, xInTopLeft,channel]
+                        if (TopRightIsIn): inTopRight = data[i,yInTopLeft, xInTopLeft+1, channel]
+                        if (BottomLeftIsIn): inBottomLeft = data[i,yInTopLeft+1, xInTopLeft, channel]
+                        if (BottomRightIsIn): inBottomRight = data[i,yInTopLeft+1, xInTopLeft+1, channel]
+
+                        out[i,yout,xout,channel] = xWeightTopLeft * yWeightTopLeft * inTopLeft\
+                             +  (1-xWeightTopLeft)*yWeightTopLeft * inTopRight\
+                             +  xWeightTopLeft * (1-yWeightTopLeft) * inBottomLeft\
+                            +(1-xWeightTopLeft) * (1-yWeightTopLeft) * inBottomRight
+        return out
+
+
+    def warp_backward_numpy(out_grad, data, grid, only_grid):
+
+        data_grad = np.zeros(data.shape)
+        grid_grad = np.zeros(grid.shape)
+
+        batchsize = data.shape[0]
+        height = data.shape[1]
+        width = data.shape[2]
+        num_channel = data.shape[3]
+
+        for i in range(batchsize):
+            for yout in range(height):
+                for xout in range(width):
+
+                    x_displacement = grid[i,yout,xout,0]
+                    y_displacement = grid[i,yout,xout,1]
+
+                    xcoord = x_displacement + xout
+                    ycoord = y_displacement + yout
+
+                    xInTopLeft = int(floor(xcoord))
+                    xWeightTopLeft = max(0, 1-abs(xcoord - xInTopLeft))
+
+                    yInTopLeft = int(floor(ycoord))
+                    yWeightTopLeft = max(0, 1-abs(ycoord - yInTopLeft))
+
+                    TopLefIsIn = xInTopLeft >= 0 and xInTopLeft <= width-1 and yInTopLeft >= 0 and yInTopLeft <= height-1;
```

Comment:
this logic happens many times. try to use a macro to wrap it. for example INSIDE()